File: app/src/main/java/com/chichiangho/base/utils/PreferenceUtil.java
package com.chichiangho.base.utils;

import android.content.Context;
import android.content.SharedPreferences;

import com.chichiangho.base.base.BaseApplication;
import com.chichiangho.base.base.ProjectDefines;
import com.chichiangho.base.bean.User;

public class PreferenceUtil {

    private static User user;

    public static User getUser() {
        if (user == null) {
            String userJson = (String) get(ProjectDefines.PREFERENCE_KEY_USER_INFO, "");
            if (!userJson.equals(""))
                return JsonUtil.fromJson(userJson, User.class);
        }
        return user;
    }

    public static void setUser(User user) {
        if (user == null)
            return;

        PreferenceUtil.user = user;
        set(ProjectDefines.PREFERENCE_KEY_USER_INFO, JsonUtil.toJson(user));
    }

    public static void set(String key, Object data) {
        set(key, data, true);
    }

    public static void set(String key, Object data, boolean asynchronous) {
        String type = data.getClass().getSimpleName();
        SharedPreferences sharedPreferences = BaseApplication.getInstance()
                .getSharedPreferences(ProjectDefines.PREFERENCE_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();

        if ("Integer".equals(type)) {
            editor.putInt(key, (Integer) data);
        } else if ("Boolean".equals(type)) {
            editor.putBoolean(key, (Boolean) data);
        } else if ("String".equals(type)) {
            editor.putString(key, (String) data);
        } else if ("Float".equals(type)) {
            editor.putFloat(key, (Float) data);
        } else if ("Long".equals(type)) {
            editor.putLong(key, (Long) data);
        }

        if (asynchronous)
            editor.apply();//异步方式写入磁盘，如果对缓存数据正确性,实时保存要求很高应使用commit()
        else
            editor.commit();
    }

    public static Object get(String key, Object defValue) {
        String type = defValue.getClass().getSimpleName();
        SharedPreferences sharedPreferences = BaseApplication.getInstance().getSharedPreferences
                (ProjectDefines.PREFERENCE_NAME, Context.MODE_PRIVATE);

        if ("Integer".equals(type)) {
            return sharedPreferences.getInt(key, (Integer) defValue);
        } else if ("Boolean".equals(type)) {
            return sharedPreferences.getBoolean(key, (Boolean) defValue);
        } else if ("String".equals(type)) {
            return sharedPreferences.getString(key, (String) defValue);
        } else if ("Float".equals(type)) {
            return sharedPreferences.getFloat(key, (Float) defValue);
        } else if ("Long".equals(type)) {
            return sharedPreferences.getLong(key, (Long) defValue);
        }

        return defValue;
    }

}

File: app/src/main/java/com/chichiangho/base/utils/PreferenceUtil.kt
package com.chichiangho.base.utils

import android.content.Context
import com.chichiangho.base.base.BaseApplication
import com.chichiangho.base.base.ProjectDefines
import com.chichiangho.base.bean.User

object PreferenceUtil {

    private var user: User? = null

    fun getUser(): User? {
        if (user == null) {
            val userJson = get(ProjectDefines.PREFERENCE_KEY_USER_INFO, "") as String
            if (userJson != "")
                return JsonUtil.fromJson(userJson, User::class.java)
        }
        return user
    }

    fun setUser(user: User?) {
        if (user == null)
            return

        PreferenceUtil.user = user
        set(ProjectDefines.PREFERENCE_KEY_USER_INFO, JsonUtil.toJson(user))
    }

    @JvmOverloads fun set(key: String, data: Any, asynchronous: Boolean = true) {
        val type = data.javaClass.simpleName
        val sharedPreferences = BaseApplication.getInstance()
                .getSharedPreferences(ProjectDefines.PREFERENCE_NAME, Context.MODE_PRIVATE)
        val editor = sharedPreferences.edit()

        if ("Integer" == type) {
            editor.putInt(key, data as Int)
        } else if ("Boolean" == type) {
            editor.putBoolean(key, data as Boolean)
        } else if ("String" == type) {
            editor.putString(key, data as String)
        } else if ("Float" == type) {
            editor.putFloat(key, data as Float)
        } else if ("Long" == type) {
            editor.putLong(key, data as Long)
        }

        if (asynchronous)
            editor.apply()//异步方式写入磁盘，如果对缓存数据正确性,实时保存要求很高应使用commit()
        else
            editor.commit()
    }

    operator fun get(key: String, defValue: Any): Any {
        val type = defValue.javaClass.simpleName
        val sharedPreferences = BaseApplication.getInstance().getSharedPreferences(ProjectDefines.PREFERENCE_NAME, Context.MODE_PRIVATE)

        if ("Integer" == type) {
            return sharedPreferences.getInt(key, defValue as Int)
        } else if ("Boolean" == type) {
            return sharedPreferences.getBoolean(key, defValue as Boolean)
        } else if ("String" == type) {
            return sharedPreferences.getString(key, defValue as String)
        } else if ("Float" == type) {
            return sharedPreferences.getFloat(key, defValue as Float)
        } else if ("Long" == type) {
            return sharedPreferences.getLong(key, defValue as Long)
        }

        return defValue
    }

}
